contracts/external/lib/Bytes.sol

```diff
@@ -0,0 +1,196 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+pragma solidity 0.5.7;
+pragma experimental ABIEncoderV2;
+
+import { Require } from "../../protocol/lib/Require.sol";
+
+
+/**
+ * @title Bytes
+ * @author dYdX
+ *
+ * Library to manipulate bytes
+ */
+library Bytes {
+    // ============ Constants ============
+
+    bytes32 constant FILE = "Bytes";
+
+    // ============ Functions ============
+
+    /**
+     * Gets the memory address for the contents of a byte array.
+     *
+     * @param  input  Byte array to lookup.
+     * @return        Memory address of the contents of the byte array.
+     */
+    /* solium-disable-next-line security/no-named-returns */
+    function contentAddress(bytes memory input)
+        internal
+        pure
+        returns (uint256 memoryAddress)
+    {
+        /* solium-disable-next-line security/no-inline-assembly */
+        assembly {
+            memoryAddress := add(input, 32)
+        }
+        return memoryAddress;
+    }
+
+    /**
+     * Writes a bytes32 into a specific position in a byte array.
+     *
+     * @param  b      Byte array to insert <input> into.
+     * @param  index  Index in byte array of <input>.
+     * @param  input  bytes32 to put into byte array.
+     */
+    function writeBytes32(
+        bytes memory b,
+        uint256 index,
+        bytes32 input
+    )
+        internal
+        pure
+    {
+        Require.that(
+            b.length >= index + 32,
+            FILE,
+            "Cannot write 32 bytes"
+        );
+
+        /* solium-disable-next-line security/no-inline-assembly */
+        assembly {
+            mstore(add(add(b, 32), index), input)
+        }
+    }
+
+    /**
+     * Copies `length` bytes from memory location `source` to `dest`.
+     *
+     * @param  dest    memory address to copy bytes to.
+     * @param  source  memory address to copy bytes from.
+     * @param  length  number of bytes to copy.
+     */
+    function memCopy(
+        uint256 dest,
+        uint256 source,
+        uint256 length
+    )
+        internal
+        pure
+    {
+        if (length < 32) {
+            // Handle a partial word by reading destination and masking
+            // off the bits we are interested in.
+            // This correctly handles overlap, zero lengths and source == dest
+            /* solium-disable-next-line security/no-inline-assembly */
+            assembly {
+                let mask := sub(exp(256, sub(32, length)), 1)
+                let s := and(mload(source), not(mask))
+                let d := and(mload(dest), mask)
+                mstore(dest, or(s, d))
+            }
+        } else {
+            // Skip the O(length) loop when source == dest.
+            if (source == dest) {
```

Comment:
why would you ever call with `source == dest`? shouldn't that be an assert error, seems like it could lead to weird behavior

Reply:
removed function